alf/algorithms/ppo_algorithm.py

```diff
@@ -52,23 +52,30 @@ def train_step(self, inputs: TimeStep, state, rollout_info):
                 reward=alg_step.info.reward,
                 discount=alg_step.info.discount,
                 action_distribution=alg_step.info.action_distribution,
-                value=alg_step.info.value))
+                value=alg_step.info.value,
+                reward_weights=alg_step.info.reward_weights))
 
     def preprocess_experience(self, root_inputs: TimeStep, rollout_info,
                               batch_info):
         """Compute advantages and put it into exp.rollout_info."""
+
+        if rollout_info.reward.ndim == 3:
+            # [B, T, D] or [B, T, 1]
+            discounts = rollout_info.discount.unsqueeze(-1) * self._loss.gamma
+        else:
+            # [B, T]
+            discounts = rollout_info.discount * self._loss.gamma
+
         advantages = value_ops.generalized_advantage_estimation(
             rewards=rollout_info.reward,
             values=rollout_info.value,
             step_types=rollout_info.step_type,
-            discounts=rollout_info.discount * self._loss._gamma,
+            discounts=discounts,
             td_lambda=self._loss._lambda,
             time_major=False)
-        advantages = torch.cat([
-            advantages,
-            torch.zeros(*advantages.shape[:-1], 1, dtype=advantages.dtype)
-        ],
-                               dim=-1)
+        advantages = tensor_utils.tensor_extend_zero(
+            advantages.transpose(0, 1)).transpose(0, 1)
```

Comment:
If you want to change to use tensor_extend_zero, I feel it's better to enchance tensor_extend_zero to add a dim argument.